x/fees/ante/eth.go

```diff
@@ -0,0 +1,64 @@
+package ante
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
+)
+
+// EthMinPriceFeeDecorator will check if the transaction's fee is at least as large
+// as the MinGasPrices param. If fee is too low, decorator returns error and tx
+// is rejected. This applies for both CheckTx and DeliverTx. This applies regardless
+// if london fork or feemarket are enabled
+// If fee is high enough, then call next AnteHandler
+type EthMinPriceFeeDecorator struct {
+	feesKeeper FeesKeeper
+	evmKeeper  EvmKeeper
+}
+
+func NewEthMinPriceFeeDecorator(fk FeesKeeper, ek EvmKeeper) EthMinPriceFeeDecorator {
+	return EthMinPriceFeeDecorator{feesKeeper: fk, evmKeeper: ek}
+}
+
+func (empd EthMinPriceFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	evmDenom := empd.evmKeeper.GetParams(ctx).EvmDenom
+	minGasPrice := empd.feesKeeper.GetParams(ctx).MinGasPrice
+	minGasPrices := sdk.DecCoins{sdk.DecCoin{
+		Denom:  evmDenom,
+		Amount: minGasPrice,
+	}}
+
+	if !minGasPrices.IsZero() {
+		for _, msg := range tx.GetMsgs() {
+			ethMsg, ok := msg.(*evmtypes.MsgEthereumTx)
+			if !ok {
+				return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid message type %T, expected %T", msg, (*evmtypes.MsgEthereumTx)(nil))
+			}
+
+			feeAmt := ethMsg.GetFee()
+
+			// For dynamic transactions, GetFee() uses the GasFeeCap value, which
+			// is the maximum gas price that the signer can pay. In practice, the
+			// signer can pay less, if the block's base fee is lower. In this case,
+			// we use the effective price
+			txData, err := evmtypes.UnpackTxData(ethMsg.Data)
+			if err == nil && txData.TxType() != ethtypes.LegacyTxType {
+				paramsEvm := empd.evmKeeper.GetParams(ctx)
+				ethCfg := paramsEvm.ChainConfig.EthereumConfig(empd.evmKeeper.ChainID())
+				baseFee := empd.evmKeeper.GetBaseFee(ctx, ethCfg)
+				feeAmt = ethMsg.GetEffectiveFee(baseFee)
+			}
```

Comment:
@fedekunze  I am using `GetEffectiveFee` for dynamic transactions, to avoid loopholes when a user can use a high gas fee cap, but a lower base fee results in an effective price < `MinGasPrice`. So, `MinGasPrice` has priority over the `feemarket` formula.

Reply:
could it be the case if the `MinGasPrice * GasLimit` is higher than the effective fee that we end up in a "deadlock" situation where txs can't be processed?

Reply:
> could it be the case if the MinGasPrice * GasLimit is higher than the effective fee that we end up in a "deadlock" situation where txs can't be processed?

If users increase the `GasTipCap` enough, the effective fee increases and transactions will pass. Tested this, as mentioned in a comment.

> reject transactions with MinGasPrices < fees < BaseFee

Actually, it should be `reject transactions with MinGasPrices * gasUsed < tx fees < EffectiveFee`. I corrected several test descriptions that were inaccurate. And added a better explanation in the AnteHandle.